src/generic/stage2/structural_iterator.h

```diff
@@ -2,29 +2,34 @@ namespace stage2 {
 
 class structural_iterator {
 public:
-  really_inline structural_iterator(const uint8_t* _buf, size_t _len, const uint32_t *_structural_indexes, size_t next_structural_index)
-    : buf{_buf},
-     len{_len},
-     structural_indexes{_structural_indexes},
-     next_structural{next_structural_index}
-    {}
-  really_inline char advance_char() {
-    idx = structural_indexes[next_structural];
-    next_structural++;
-    c = *current();
-    return c;
+  const uint8_t* const buf;
+  uint32_t *current_structural;
+  dom_parser_implementation &parser;
+
+  // Start a structural 
+  really_inline structural_iterator(dom_parser_implementation &_parser, size_t start_structural_index)
+    : buf{_parser.buf},
+      current_structural{&_parser.structural_indexes[start_structural_index]},
+      parser{_parser} {
+  }
+  // Get the buffer position of the current structural character
+  really_inline const uint8_t* current() {
+    return &buf[*current_structural];
   }
+  // Get the current structural character
   really_inline char current_char() {
-    return c;
+    return buf[*current_structural];
   }
+  // Get the next structural character without advancing
```

Comment:
I don't care, but shouldn't this be renamed to `peek_next_char` then?

Reply:
I would also like to point out that this function is barely used at all... and could almost be replaced by "is_next_char_an_open_bracket". :-)

Reply:
Fixed :) I think that makes sense. I didn't change it to is_next_char_an_open_bracket because I wanted to keep knowledge about *which* characters are in the buffer out of the structural_iterator (cleaner separation of purpose, it just iterates and gives you characters/buffers).